File: packages/plugins/notifications/Sources/TanoPluginNotifications/NotificationsPlugin.swift
import Foundation
import TanoBridge

#if canImport(UserNotifications)
import UserNotifications
#endif

/// Native local-notifications plugin for Tano.
///
/// Provides permission requests, scheduling, and cancellation of local
/// notifications via UNUserNotificationCenter on iOS.
/// On macOS (test environment), returns success stubs.
///
/// Supported methods: `requestPermission`, `schedule`, `cancel`.
public final class NotificationsPlugin: TanoPlugin {

    // MARK: - TanoPlugin conformance

    public static let name = "notifications"
    public static let permissions: [String] = ["notifications"]

    public init() {}

    // MARK: - Routing

    public func handle(method: String, params: [String: Any]) async throws -> Any? {
        switch method {
        case "requestPermission":
            return try await requestPermission()
        case "schedule":
            return try await schedule(params: params)
        case "cancel":
            return try await cancel(params: params)
        default:
            throw NotificationsPluginError.unknownMethod(method)
        }
    }

    // MARK: - requestPermission

    private func requestPermission() async throws -> [String: Any] {
        #if os(iOS)
        let center = UNUserNotificationCenter.current()
        let granted = try await center.requestAuthorization(options: [.alert, .sound, .badge])
        return ["granted": granted]
        #else
        // Stub for macOS tests
        return ["granted": true]
        #endif
    }

    // MARK: - schedule

    private func schedule(params: [String: Any]) async throws -> [String: Any] {
        guard let title = params["title"] as? String else {
            throw NotificationsPluginError.missingParameter("title")
        }

        let body = params["body"] as? String ?? ""
        let delay: TimeInterval
        if let d = params["delay"] as? Double {
            delay = d
        } else if let d = params["delay"] as? Int {
            delay = TimeInterval(d)
        } else {
            delay = 1.0
        }

        let identifier = UUID().uuidString

        #if os(iOS)
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default

        let trigger = UNTimeIntervalNotificationTrigger(
            timeInterval: max(delay, 1),
            repeats: false
        )

        let request = UNNotificationRequest(
            identifier: identifier,
            content: content,
            trigger: trigger
        )

        try await UNUserNotificationCenter.current().add(request)
        #endif

        return ["id": identifier]
    }

    // MARK: - cancel

    private func cancel(params: [String: Any]) async throws -> [String: Any] {
        guard let identifier = params["id"] as? String else {
            throw NotificationsPluginError.missingParameter("id")
        }

        #if os(iOS)
        UNUserNotificationCenter.current()
            .removePendingNotificationRequests(withIdentifiers: [identifier])
        #endif

        return ["success": true]
    }
}

// MARK: - Errors

public enum NotificationsPluginError: Error, LocalizedError {
    case unknownMethod(String)
    case missingParameter(String)

    public var errorDescription: String? {
        switch self {
        case .unknownMethod(let m):
            return "Unknown notifications plugin method: \(m)"
        case .missingParameter(let p):
            return "Missing required parameter: \(p)"
        }
    }
}

File: packages/android/src/main/kotlin/dev/tano/plugins/NotificationsPlugin.kt
package dev.tano.plugins

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationCompat
import dev.tano.bridge.TanoPlugin
import java.util.UUID

/**
 * Native local-notifications plugin for Tano on Android.
 *
 * Provides permission requests, scheduling, and cancellation of local
 * notifications via [NotificationManager] and [NotificationChannel].
 *
 * Supported methods: `requestPermission`, `schedule`, `cancel`.
 *
 * Mirrors the iOS NotificationsPlugin.
 *
 * @param context Android Context needed to access NotificationManager.
 */
class NotificationsPlugin(private val context: Context) : TanoPlugin {

    companion object {
        private const val CHANNEL_ID = "dev.tano.notifications"
        private const val CHANNEL_NAME = "Tano Notifications"
    }

    override val name: String = "notifications"
    override val permissions: List<String> = listOf("notifications")

    private val notificationManager: NotificationManager
        get() = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    // Track notification IDs: string identifier -> int id
    private val notificationIds = mutableMapOf<String, Int>()
    private var nextId = 1

    // -- Routing --

    override suspend fun handle(method: String, params: Map<String, Any?>): Any? {
        return when (method) {
            "requestPermission" -> requestPermission()
            "schedule" -> schedule(params)
            "cancel" -> cancel(params)
            else -> throw IllegalArgumentException("Unknown notifications plugin method: $method")
        }
    }

    // -- requestPermission --

    private fun requestPermission(): Map<String, Any?> {
        // On Android 8.0+, create the notification channel (required).
        // On Android 13+ (API 33), POST_NOTIFICATIONS runtime permission is
        // needed but must be requested from an Activity; here we just ensure
        // the channel exists and report granted=true optimistically.
        ensureChannel()
        return mapOf("granted" to true)
    }

    // -- schedule --

    private fun schedule(params: Map<String, Any?>): Map<String, Any?> {
        val title = params["title"] as? String
            ?: throw IllegalArgumentException("Missing required parameter: title")
        val body = params["body"] as? String ?: ""
        val delay = when (val d = params["delay"]) {
            is Double -> d
            is Int -> d.toDouble()
            is Long -> d.toDouble()
            else -> 1.0
        }

        ensureChannel()

        val identifier = UUID.randomUUID().toString()
        val notifId = synchronized(notificationIds) {
            val id = nextId++
            notificationIds[identifier] = id
            id
        }

        // For simplicity, post the notification immediately.
        // A production implementation would use AlarmManager or WorkManager
        // for delayed delivery.
        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setContentTitle(title)
            .setContentText(body)
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .setAutoCancel(true)
            .build()

        notificationManager.notify(notifId, notification)

        return mapOf("id" to identifier)
    }

    // -- cancel --

    private fun cancel(params: Map<String, Any?>): Map<String, Any?> {
        val identifier = params["id"] as? String
            ?: throw IllegalArgumentException("Missing required parameter: id")

        val notifId = synchronized(notificationIds) {
            notificationIds.remove(identifier)
        }

        if (notifId != null) {
            notificationManager.cancel(notifId)
        }

        return mapOf("success" to true)
    }

    // -- Helpers --

    private fun ensureChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_DEFAULT
            )
            notificationManager.createNotificationChannel(channel)
        }
    }
}
